rfcs/87-feature-gates.md

```diff
@@ -0,0 +1,117 @@
+---
+author: Xuanwo <github@xuanwo.io>
+status: draft
+updated_at: 2021-05-27
+---
+
+# GSP-87: Feature Gates
+
+## Background
+
+Behavior consistence is the most important thing for go-storage. However, we do have different capabilities and limitations in storage services. So the problem comes into how to handle them.
+
+Our goals are:
+
+- Behavior consistent by default (invalid operations should be return error)
+- Give user abilities to loose the restriction
+- Allow user to enable the features they want
+
+In [GSP-16], we introduced loose mode which is a global flag that controls the behavior when services meet the unsupported pairs.
+
+- If `loose` is on: This error will be ignored.
+- If `loose` is off: Storager returns a compatibility related error.
+
+However, we removed `loose` in [GSP-20] because we can't figure out [error could be returned too early while in loose mode](https://github.com/beyondstorage/go-storage/issues/233). And loose is so general that it affects nearly all behavior of Storager.
+
+In [types: Implement pair policy](https://github.com/beyondstorage/go-storage/pull/453), we try to figure out this problem by introducing `PairPolicy`.
+
+`PairPolicy` controls the behavior of pairs:
+
+```go
+type PairPolicy struct {
+	All bool
+	
+	...
+
+	// pairs for interface Storager
+	Create           bool
+	Delete           bool
+	List             bool
+	ListListMode     bool
+	Metadata         bool
+	Read             bool
+	ReadSize         bool
+	ReadOffset       bool
+	ReadIoCallback   bool
+	Stat             bool
+	Write            bool
+	WriteContentType bool
+	WriteContentMd5  bool
+	WriteIoCallback  bool
+}
+```
+
+If `PairPolicy.Write` is on, Storager will return `services.PairUnsupportedError` while meeting not supported pairs.
+
+But it doesn't solve the problem:
+
+- `PairPolicy` is generated in go-storage and can't reflect the capabilities in service.
+- `PairPolicy` only used for pair capabilities check, and can't fix the problem introduced in [GSP-86]
+
+## Proposal
+
+So I propose to treat loose behavior consistence as a feature, and introduce feature gates in [go-storage]. 
+
+### New Feature: Loose Operation
+
+By default, [go-storage] will return errors for not supported pairs. If loose operation feature has been enabled, [go-storage] will ignore those errors.
+
+To enable loose operation, users need to add pairs like:
+
+```go
+s3.WithStorageFeatures(s3.StorageFeatures{
+	LooseOpeartionAll: true,
+	LooseOperationWrite: true,
+})
+```
+
+> `New` function is special, and we will always enable `loose` for it.
+
+### New Feature: Virtual Operation
```

Comment:
We will introduce a new idea `virtual` to represents following state: A service doesn't support some feature, but we can simulate it by some methods.

For example:

- We can have `Virtual Operation`
  - S3 doesn't have native support for `CreateDir`, but we can put an Object end with `/` to simulate it.
  - S3 doesn't have native support for `Link`, but we can store the link target in object metadata to simulate it.
- We can have `Virtual Pair`
  - fs doesn't support `content_md5` for `write`, but we can store it in file's xattr to simulate it.

Those features will affect users data in some way:

- User couldn't read those data without go-storage
- Those data could be changed by other API and go-storage can't detect them.

So the end user have to decide whether enable those virtual features or not.

- If they enabled, our services will run in the `virtual` mode, likes your virtual machine. And they have to afford the side effects.
- If they don't, our services will behave like they don't implement this feature.

The virtual feature will give us more power so that we can implement `Link` / `Rename` even when our service doesn't have native support.